Source/AwsCommonRuntimeKit/crt/Future.swift

```diff
@@ -0,0 +1,78 @@
+//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//  SPDX-License-Identifier: Apache-2.0.
+
+import Foundation
+
+public final class Future<Value> {
+    public typealias FutureResult = Result<Value, Error>
+
+    private var _value: FutureResult? //nil when pending
+
+    private var waiter = DispatchSemaphore(value: 0)
+
+    private var _observers: [((FutureResult) -> Void)] = []
+
+    public init(value: FutureResult? = nil) {
+        self._value = value
+    }
+
+    public func get() throws -> Value {
+        if let value = self._value {
```

Comment:
Does this need to be an atomic load? If this were C/C++ it would have to be.

Reply:
I think I handled this with a lock everywhere and unit tests frmo multiple threads now pass.